src/hardware/mixer.cpp

```diff
@@ -535,17 +542,130 @@ constexpr void fill_8to16_lut()
 		lut_u8to16[i] = u8to16(i);
 }
 
+template <class Type, bool stereo, bool signeddata, bool nativeorder>
+AudioFrame MixerChannel::ConvertNextFrame(const Type *data, const work_index_t pos)
+{
```

Comment:
This is a very nice and significant refactor!

Would be really nice to wrap some unit tests around this that hit it with the handful of template variations, in-and-out of bounds values, just as a sanity check (and probably add asserts in here too, when we know the data should be bounded).

Reply:
If adding the tests, suggest inserting that commit prior to this refactor, so the baseline code gets tested/proven first, and then the tests will carry forward into this refactoring commit - and given the API is the same, will be very nice to see all those passing too :+1:

Reply:
Yeah, could be done.